Block Bindings: Allow editing in post meta source

SantosGuillamot opened the pull request:
## What?
Use the latest changes to the block bindings API in the editor to allow users to edit the value of the custom fields directly through the blocks when they are connected. For example, when the paragraph content is bound to a specific custom field, and the user start typing, they would be modifying the custom field and not the content.

At least in this first iteration, it should be limited to:
* Users with the permissions to edit the field.
* When users are editing a post, and not a template.
* Custom fields should not be protected and available through the REST API.

## How?
Using the `setValue` function based on the code created by @ellatrix  [here](https://github.com/WordPress/gutenberg/pull/60725). Apart from that, we should use the `lockAttributesEditing` callback added in this pull request to cover the mentioned limitation.

Additionally, it ensure that the split selection work as expected in bound blocks.

## Testing Instructions
I added a couple of e2e tests for this use case. You can test it manually by:

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
<details>
<summary>1. Register post meta by adding this snippet to your theme's functions.php</summary>

```php
add_action( 'init', 'test_block_bindings' );

function test_block_bindings() {
	register_meta(
		'post',
		'text_field',
		array(
			'show_in_rest'      => true,
			'single'            => true,
			'type'              => 'string',
			'default'           => 'default text value',
		)
	);
}
```

</details>


<details>

<summary>2. Add a paragraph block bound to the custom field using the Code Editor</summary>

```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"text_field"}}}}} -->
<p>Paragraph content</p>
<!-- /wp:paragraph -->
```

</details>
 
3. Check that the value of the custom field is shown.
4. Check you can edit it when you are in the post editor.
5. Check that modifying the value actually changes the custom field and not the paragraph content.

SantosGuillamot commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -25,4 +25,25 @@ export default {
+
+		// TODO: Check that the custom field is not protected and available in the REST API.
+
+		// TODO: Check that the user has the capability to edit post meta.
```
@cbravobernal explored the possibility of creating a new selector to check if the user is an admin [here](https://github.com/WordPress/gutenberg/pull/61734/commits/9659455eed4343c455760b2f9076fe3a5bdbf2d7). We also discussed the possibility of using [canUserEditEntityRecord](https://github.com/WordPress/gutenberg/blob/9659455eed4343c455760b2f9076fe3a5bdbf2d7/packages/core-data/src/selectors.ts#L1164).

gziolo commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -25,4 +25,25 @@ export default {
+			: registry.select( editorStore ).getCurrentPostType();
+		registry
+			.dispatch( coreDataStore )
+			.editEntityRecord( 'postType', postType, context.postId, {
```
Should there be a callback for the case when `postId` is not set on the context? Similar to the `postType`?

SantosGuillamot replied:
I've added the fallback for the `postId` as well. Although thinking a bit more about it I am not sure if we should include fallback at all for both the `postType` and `postId` or just rely on them being exposed through `usesContext`.

SantosGuillamot commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -182,11 +182,14 @@ export function RichTextWrapper(
 						! blockBindingsSource ||
-						blockBindingsSource.lockAttributesEditing
+						blockBindingsSource.lockAttributesEditing( {
+							select,
+							args: binding.args,
```
I am not sure how to pass the context from the rich-text.

SantosGuillamot commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -25,4 +25,48 @@ export default {
+				},
+			} );
+	},
+	lockAttributesEditing( { select, context = {}, args } ) {
```
It seems we will need these arguments for the `lockAttributesEditing` callback as well.

I don't know if we should pass the registry like in the other functions or the `select` is enough.

Any thoughts?

cbravobernal replied:
I'm usually in favor on keeping consistency with other functions.

SantosGuillamot replied:
In [this other pull request](https://github.com/WordPress/gutenberg/pull/61817/files#diff-cf4447b4134dbf24fa2819703999f0131449fb79a0611bd544c5e05fd80aa36bR19), I'm changing `getValue` to receive `select`. I believe users shouldn't use the whole register in the `getValue` function.

And I believe it is the same case for locking the attributes editing, so I think it is fine to have it here as well.

SantosGuillamot commented:
There is an issue when pressing "Enter" to add a new block, and the rich text is not disabled. I'll take a look at that.

cbravobernal commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -16,13 +16,64 @@ export default {
+		}
+
+		// Check that the user has the capability to edit post meta.
+		const canUserEdit = select( coreDataStore ).canUserEditEntityRecord(
+			'postType',
+			postType,
+			postId
+		);
+		if ( ! canUserEdit ) {
+			return true;
+		}
+
+		return false;
```
I've been thinking about this and, if your are a contributor or an author user. You are not even allowed to access to the post editor part. So does it make sense to have this check? 

I think it will be always true (this is only called in the post editor page, meaning that the user is an author/contributor of their own posts, an editor or an admin, canUserEditEntityRecord is true in all of those cases)

![Screenshot 2024-05-17 at 16 57 44](https://github.com/WordPress/gutenberg/assets/37012961/81d32162-2d6a-4187-b7cf-27823c674263)


Auto answer: I'm thinking about the case of creating a custom post meta field, and then only allow certain roles to edit them, but it would be a custom solution, so maybe it fits better on each source rather than post-meta 🤔

SantosGuillamot replied:
> So does it make sense to have this check?

I was wondering the same, but I left it just in case. @gziolo Any idea if this is necessary at all?

gziolo replied:
What if you are editing a post as an author and there is a custom block that sets the post id in the context to a different value in all inner blocks. That potentially could be also a Query Loop that has some core blocks with connected attributes. As an author you shouldn’t be able to edit anything inside these nested connected block in the posts you don’t authored. It’s just a single example but options  are endless with custom blocks. Another example would be a feature product block that allows customizations in inner blocks. A different set of challenges will start to arise with more sources. How do you prevent editing the site title, site logo or site tagline when injected in the post content?

SantosGuillamot replied:
Thanks for the examples. I think they totally make sense. It seems it is worth keeping the `canUserEditEntityRecord`.

I assume that for sources like site data, we could still rely on similar methods to `canUser` and `canUserEditEntityRecord`.

cbravobernal commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -16,13 +16,64 @@ export default {
+	},
+	lockAttributesEditing( { select, context = {}, args } ) {
+		const postId = context.postId
+			? context.postId
```
Is there any special reason of initialising the context as an object rather than doing `context?.postId` ?

SantosGuillamot replied:
No reason, I can change that 🙂 

EDIT: Done [here](https://github.com/WordPress/gutenberg/pull/61753/commits/f66fb79aef987b892463b2654b5aa02ed2ab3ec3).

SantosGuillamot commented:
> There is an issue when pressing "Enter" to add a new block, and the rich text is not disabled.

I think the root of the problem is that when we call `getBlockAttributes` it returns the value of the original content and not the one modified with the custom field value. For this reason, when we call it [here](https://github.com/WordPress/gutenberg/blob/995607a169577432046e583eea61625a8571800b/packages/block-editor/src/store/actions.js#L921-L923) during `__unstableSplitSelection` function, it is not correct and it causes some issues.

Not sure if modifying the attributes in the store itself instead of using the current hook would help in this regard.

SantosGuillamot commented:
I have rebased this branch into [this other pull request](https://github.com/WordPress/gutenberg/pull/61817) because it is supposed to solve the issues I was encountering while pressing Enter on bound blocks.

gziolo commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -28,4 +28,49 @@ export default {
+				},
+			} );
+	},
+	lockAttributesEditing( { select, context, args } ) {
```
I was a bit confused when I saw this method name as I expected it's an action that locks attribute editing. What alternatives would help here to better express intent – check whether the connected value can be updated?

SantosGuillamot replied:
You're right it can be confusing. We could change the intent to just the opposite. Some ideas that come to my mind:
* `allowEditing`
* `canEditSource`
* `canUpdateSource`
* `canUpdateSourceValue`

Any ideas are welcome 🙂

gziolo replied:
One important consideration is that the user might be able to edit source, but not necessarily all posts or individual fields, as permissions are very granular. I don't know if that gets reflected on the client.

I see that permissions are checked only on the post level with `canUserEditEntityRecord `. In effect, `canUserEditValue` or `canEditValue` might resonate better.

SantosGuillamot replied:
I've changed the name to `canUserEditValue` in the [latest commit](https://github.com/WordPress/gutenberg/pull/61753/commits/570a4b287de28c6b95a55a1ee0e78781c960054b). It looks better to me. Anyways, the API is still private so we can change that if wanted.

> One important consideration is that the user might be able to edit source, but not necessarily all posts or individual fields, as permissions are very granular. I don't know if that gets reflected on the client.

Do you mean that the current conditions might not be enough? We are assuming that if these conditions are `true`, the user can edit the meta field:
* The user can edit the current post.
* The meta field is publicly exposed in the REST API.

What else should we consider?

gziolo replied:
I was thinking about the [`is_protected_meta`](https://developer.wordpress.org/reference/functions/is_protected_meta/) function that allows filtering, but you are correct that when the individual meta fields gets exposed through REST API for the current user (and their role) then we can safely assume they can edit it 👍🏻

SantosGuillamot commented:
I pushed some changes trying a new approach to solve [the issue we were having with the "enter"](https://github.com/WordPress/gutenberg/pull/61753#issuecomment-2118008637). This would allow us merge this pull request without having to [expose the context in the store](https://github.com/WordPress/gutenberg/pull/61895) as it is not clear yet if that should be the approach or its implications.

However, I had some issues with this approach as well. Let me try to explain what I did:

* First, I've created a [new branch](https://github.com/WordPress/gutenberg/pull/62063) to create a util to transform the block attributes with binding values. This pull request is on top of that one.
* Then, I'm using that helper to use the transformed attributes instead of the fallback: [link](https://github.com/WordPress/gutenberg/pull/61753/files#diff-d1fbd3af72ee0220c5c3ab2952a24d818174aed711b9a50b57c9544218a63988R923-R930).
* For doing that, I needed to pass the context to the `__unstableSplitSelection` function in `use-input`: [link](https://github.com/WordPress/gutenberg/pull/61753/files#diff-72ac86ae2dfc766e44b27373d3d046a73bfaa72c46677f6069422d62840b120eR121).

This last point me was my main concern, but I couldn't find a way to do it without passing the context (or exposing it in the store). Any thoughts?

cc: @gziolo @artemiomorales

artemiomorales commented:
@SantosGuillamot I'm going to look more in depth at the code here to see if I can get to the point where I can offer an opinion on your questions, but for now I'll just say that I tested this code and it works as intended 👍

I also combined it with https://github.com/WordPress/gutenberg/pull/61811 locally and they seem to be working as intended together without issues.

ellatrix commented on packages/block-editor/src/store/actions.js:
```diff
@@ -918,9 +920,18 @@ export const __unstableSplitSelection =
+					selectionA.clientId,
+					context,
+					registry
+				);
```
Shouldn't we disable splitting and merging when an attribute is bound? It doesn't make much sense imo. Binding a paragraph makes it a single entity.

SantosGuillamot replied:
Maybe that's another possibility. What should happen when you are editing a bound heading and you press "Enter"?

SantosGuillamot replied:
Maybe we can just insert a new block when pressing Enter in a bound block. Do you think that would be better? Any idea how to do that?

gziolo replied:
I think that would be a good first step to disable splitting and merging when in the context of bound attributes. I can't think of a case where someone would intend to create a new block that carries over some part of the edited value to the new block.

In the long run, I would even consider making the more explicit distinction in the user experience that for bound values in RichText you are only editing an external value and shouldn't expect it works the same as with regular attributes.

gziolo replied:
> First, I've created a https://github.com/WordPress/gutenberg/pull/62063 to create a util to transform the block attributes with binding values. This pull request is on top of that one.
> Then, I'm using that helper to use the transformed attributes instead of the fallback: [link](https://github.com/WordPress/gutenberg/pull/61753/files#diff-d1fbd3af72ee0220c5c3ab2952a24d818174aed711b9a50b57c9544218a63988R923-R930).

Assuming you only check whether the attribute has a bound value, you could avoid that logic, and assume that `enter` finishes editing and create a new block with no content inside. That would be similar to the current behavior when handling the value in the readonly mode.

SantosGuillamot replied:
I've just tried this approach in [this commit](https://github.com/WordPress/gutenberg/pull/61753/commits/448623294bcbe7047fea7edefcfbc05ea0bc3500#diff-d1fbd3af72ee0220c5c3ab2952a24d818174aed711b9a50b57c9544218a63988R875-R884), and it seems to work. That would avoid passing context to the existing `__unstableSplitSelection` parameters.

Is that what you had in mind?

EDIT: By the way, if we follow this approach I can rebase to `trunk` and we could merge this pull request without [the other one to create a util](https://github.com/WordPress/gutenberg/pull/62063).

gziolo replied:
Yes, that's exactly what I described.

ellatrix replied:
>  What should happen when you are editing a bound heading and you press "Enter"?

Either nothing or create a soft line break (if not disabled for that rich text instance). When you're in a bound attribute it should be completely isolated imo.

SantosGuillamot commented:
Based on [this conversation](https://github.com/WordPress/gutenberg/pull/61753#discussion_r1618468993), I changed the approach of the pull request to always insert a new block when pressing "Enter" if the rich text attribute is connected. This avoids the need to pass the context to the function, and it doesn't depend on the util created in [this other pull request](https://github.com/WordPress/gutenberg/pull/62063) anymore.

I've rebased it on `trunk` and I believe it is ready for review. Happy to explore other paths if needed.

gziolo commented:
@talldan and @kevin940726, it would also be great to get a sanity check from you and ensure that everything continues to work correctly for Pattern Overrides. Currently, all e2e tests pass, but I saw some failures during the development, so better to stay on the safe side.

> Based on https://github.com/WordPress/gutenberg/pull/61753#discussion_r1618468993, I changed the approach of the pull request to always insert a new block when pressing "Enter" if the rich text attribute is connected. This avoids the need to pass the context to the function, and it doesn't depend on the util created in https://github.com/WordPress/gutenberg/pull/62063 anymore.

Code changes applied look good to me. @ellatrix, anything that raises a flag from your perspective?

I'd like to spend more time testing it as it's an impactful change from the user perspective. In the testing instructions only the example for the Paragraph block is included. What about other blocks and attributes that can be modified through different parts of UI than RichText?

ellatrix commented on packages/block-editor/src/store/actions.js:
```diff
@@ -874,6 +872,16 @@ export const __unstableSplitSelection =
+
+		// If the attribute is bound, don't split the selection and insert a new block instead.
+		if ( bindings?.[ attributeKeyA ] ) {
+			dispatch.insertAfterBlock( selectionA.clientId );
```
I'd rather have it not do anything tbh

ellatrix replied:
There's also some other nuances: what happens when you paste block into a bound attribute right now with this PR? The correct behaviour would be to paste everything inline imo (since it's an isolated editor)

SantosGuillamot replied:
> I'd rather have it not do anything tbh

I don't have a strong opinion, to be honest. I think I lean towards adding a new block but I am not sure. Maybe @jasmussen can help to decide. There are basically two options when a block is connected and users press enter:

**1. Add a new block**


https://github.com/WordPress/gutenberg/assets/34552881/db1d9396-5ffd-4fad-96e9-e8fb5a33d8d0


**2. Don't do anything**

In this case, maybe we can add a warning saying "Blocks can't be split when using bindings" or something similar.

https://github.com/WordPress/gutenberg/assets/34552881/468816aa-129b-4d78-b7b6-1fee5be69998

SantosGuillamot replied:
> There's also some other nuances: what happens when you paste block into a bound attribute right now with this PR? The correct behaviour would be to paste everything inline imo (since it's an isolated editor)

Right now, if you copy text it works as expected, but if you copy a whole block, it does nothing. I agree that probably the expected behavior is to paste everything inline. I'll try to take a look at that. Any clue where that logic lives?

SantosGuillamot replied:
I've been testing copying and pasting a block in a heading in `trunk`, and it seems to be broken, so it is hard to imagine how it should work with bindings. What I see is that after pasting a block, it creates a weird markup:

EDIT: It seems this is solved by [this pull request](https://github.com/WordPress/gutenberg/pull/61900).


https://github.com/WordPress/gutenberg/assets/34552881/d4a3c51e-1efc-4582-b205-92d70f2860cd

jasmussen replied:
Just making sure I understand: we're discussing what happens when you press Enter in a bound block, is that right?

Assuming I understand this right, I would lean towards pressing Enter making a new paragraph. I.e. not changing the writing flow behavior just for bound blocks.

SantosGuillamot replied:
For now, I added a warning in [this commit](https://github.com/WordPress/gutenberg/pull/61753/commits/2012ae84396ffb0d96829830d36483365555a733) when trying to paste a block into another one with bindings.

I believe changing the value of the meta field (or other source) after pasting might be trickier because we might need to call `setValues`, but I'll keep exploring to see if we can avoid that.

SantosGuillamot replied:
> Just making sure I understand: we're discussing what happens when you press Enter in a bound block, is that right?

Sorry I missed the comment. Yes, that was my question 🙂

SantosGuillamot replied:
Based on the feedback, I kept the logic to insert a new block when pressing Enter and not allowing to paste blocks when bindings exist until we figure out how to do it properly.

Is that okay for you, @ellatrix ? Any more concerns I should work on to get this merged?

cbravobernal replied:
I can confirm that `shift+enter` will add a line break while maintaining the text edition. 👍

cbravobernal replied:
If we want to be like actual paragraph, all the content from the cursor to the end of the paragraph should be moved to the next block.

PD: I agree that is better to do nothing if we don't have a consensus on the behavior of the enter button here. Also, is of course less prone to errors.

ellatrix replied:
My third suggestion was to insert a soft line break when splitting is not available (which is what rich text does automatically now. Tbh, I don't care that much, as long as we're not splitting the source.

SantosGuillamot commented:
I added a few more e2e tests to cover more blocks in [this commit](https://github.com/WordPress/gutenberg/pull/61753/commits/c537acbdd6b8a8ac6caff4b2a23968d733fc74f9).

SantosGuillamot committed:
Change parent context logic

SantosGuillamot committed:
Add action, selector, and reducer for block context

SantosGuillamot committed:
Go back to syncing store in edit component

SantosGuillamot committed:
Simplify fallback conditional

SantosGuillamot committed:
Create bindings util for transforming block attributes

SantosGuillamot committed:
Pass only `select` to `getValue` functions

SantosGuillamot committed:
Add conditional in block

SantosGuillamot committed:
Remove old editor hook

SantosGuillamot committed:
Change bindings destructuring

SantosGuillamot committed:
Move logic to BlockContextProvider

SantosGuillamot committed:
WIP: Move bindings logic to `getBlockAttributes`

SantosGuillamot committed:
Get attributes and name from the store

SantosGuillamot committed:
Revert "Sync store in edit component"

This reverts commit 988c4b60b9f6f0194423e19ac28e5a7ad811ea00.

SantosGuillamot committed:
Update unit tests to expect a dispatch

SantosGuillamot committed:
Add fallback to postId until context is ready

SantosGuillamot committed:
Change function to return only the bound attributes

SantosGuillamot committed:
Use useLayoutEffect

SantosGuillamot committed:
Check canBindAttribute in updateBlockAttributes

SantosGuillamot committed:
Remove setValue post-meta code

SantosGuillamot committed:
Sync store in edit component

SantosGuillamot committed:
WIP: Move bindings setAttributes logic to updateBlockBindings

SantosGuillamot committed:
Access context through selector

SantosGuillamot committed:
Don't use `getBlockAttributes` inside `getValue`

SantosGuillamot committed:
Add postType fallback when locking controls

SantosGuillamot committed:
Update tests to check contenteditable

SantosGuillamot committed:
Check user can edit post meta

SantosGuillamot committed:
Update lockAttributesEditing default

SantosGuillamot committed:
Don't use fallback for context

SantosGuillamot committed:
Test editing is allowed in paragraph block

SantosGuillamot committed:
Pass arguments to lockAttributesEditing

SantosGuillamot committed:
Disable editing in templates

SantosGuillamot committed:
Revert "Add edit value posibility for post meta, add function to check if is admin"

This reverts commit 9659455eed4343c455760b2f9076fe3a5bdbf2d7.

SantosGuillamot committed:
Pass context in rich text

SantosGuillamot committed:
Update getBlockAttributes logic

SantosGuillamot committed:
Revert "Don't use `getBlockAttributes` inside `getValue`"

This reverts commit 0e91129ee1b9fa7cf34db6567b9dd70e1d6599c6.

cbravobernal committed:
Add edit value posibility for post meta, add function to check if is admin

SantosGuillamot committed:
Revert "Revert "Enable parallel processing for PHPCS sniffs (#61700)""

This reverts commit e74d71cf0538b4aa8016b627db7849b1206e7a02.

SantosGuillamot committed:
Test protected fields are not editable

SantosGuillamot committed:
Add post meta setValue function

SantosGuillamot committed:
Check field is exposed in the REST API

SantosGuillamot committed:
Don't use fallbacks

SantosGuillamot committed:
Revert "Enable parallel processing for PHPCS sniffs (#61700)"

This reverts commit 8331820af51afb6d834b4d3e395a4adc8d993244.

SantosGuillamot committed:
Simplify lockAttributesEditing fallbacks

SantosGuillamot committed:
Adapt old tests

SantosGuillamot committed:
Avoid processing bindings recursively

SantosGuillamot committed:
Add fallback for postId

SantosGuillamot committed:
Change name to `canUserEditValue`

SantosGuillamot committed:
Pass block context through rich text

SantosGuillamot committed:
Check contenteditable attribute in test

SantosGuillamot committed:
Revert changes caused by rebase

SantosGuillamot committed:
Pass context to in use-input

SantosGuillamot committed:
Add space back

SantosGuillamot committed:
Don't split when attribute is bound

SantosGuillamot committed:
Use getBoundAttributesValues

SantosGuillamot committed:
Add a warning when pasting blocks

SantosGuillamot committed:
Change REST API check

SantosGuillamot committed:
Transform block attributes into bindings in split selection

SantosGuillamot committed:
Revert changes caused by rebasing

SantosGuillamot committed:
Cover more blocks when editing custom fields

SantosGuillamot committed:
Change imports

cbravobernal commented on packages/block-editor/src/store/actions.js:
```diff
@@ -872,6 +873,30 @@ export const __unstableSplitSelection =
+					registry.dispatch( noticesStore );
+				createWarningNotice(
+					__(
+						"Blocks can't be inserted into other blocks with bindings"
```
You can paste a paragraph inside a binded paragraph to update its content right?

I could do that on my testing, so in that case, we may change the warning to 
 `Blocks can't be insterted into different blocks with bindings`

SantosGuillamot replied:
> You can paste a paragraph inside a binded paragraph to update its content right?

Mmm, I don't think so. You can paste some text, but not if you copy the entire block I believe.

gziolo replied:
In my testing, it largely depend when you try to paste blocks. For example, try it with the Heading block, the Verse block, or caption for the Image block. It will get handled differently without any notices in the UI:


https://github.com/WordPress/gutenberg/assets/699132/249b096e-9d2f-4cff-864c-177e7014d443

In effect, my recommendation would be to take some action that fits best in the given scenario and skip the warning.

SantosGuillamot replied:
Ideally, we probably want to modify the source value with whatever is supposed to be pasted. But that seemed a bit tricky to me at this point and I thought it could be handled in a follow-up PR. That's why I added the warning in the meantime.

cbravobernal commented on packages/block-editor/src/store/actions.js:
```diff
@@ -872,6 +873,30 @@ export const __unstableSplitSelection =
+				);
+				return;
+			}
+			dispatch.insertAfterBlock( selectionA.clientId );
```
```suggestion
```
I would remove the `enter` option, is something we can add in future releases after some UX research.

cbravobernal reviewed (dismissed):
Works as expected. Apart from the `enter` behaviour. The PR is ready to land.

cbravobernal reviewed (approved):
It seems that the custom fields edit panel never worked well with bindings, so should not be a block for this PR.

ellatrix commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -16,13 +16,53 @@ export default {
+				},
+			} );
+	},
+	canUserEditValue( { select, context, args } ) {
```
This is all private right?

gziolo replied:
Yes, registering block binding sources is still private. If everything goes well in WP 6.6 cycle, we are going to open it to everyone in WP 6.7.

SantosGuillamot replied:
Yes, and we will use next cycle to polish the APIs. There are some changes I would like to explore.

gziolo commented:
It's rather unrelated to this PR, but I wanted to surface it as I noticed while testing:

<img width="875" alt="Screenshot 2024-05-31 at 13 01 22" src="https://github.com/WordPress/gutenberg/assets/699132/d75128bc-ad77-4193-9a4c-3745edcb6235">

I tried to set a binding on the `caption` attribute, but it is still not covered. However, it is listed in the sidebar.

gziolo commented:
Everything works correctly in my testing. It looks like permissions are respected, so when editing the post as an author, when I use Query block to display other posts, I can't modify meta values for fields I don't have access to:


https://github.com/WordPress/gutenberg/assets/699132/cdea7bee-6b67-47ca-abee-600098140120

Overall, I can see only two aspects to watch after this very exciting feature gets enabled for users:
- Iterate on how the pasting behavior works, but that might be a larger effort to unify it, as noted in https://github.com/WordPress/gutenberg/pull/61753#discussion_r1622232372.
- "There is an issue when pressing "Enter" to add a new block, and the rich text is not disabled." - that issue seems to be resolved and works pretty decent. The only part that I would pay attention to is how users feel about the fact that pressing `enter` at any position in the edited text when using RichText creates a new block and moves focus there. It doesn't seem that different to other flows I tested, like when typing an URL, or renaming a block title in the modal triggered from the list view.

gziolo reviewed (approved):
Excellent work bringing it to the finish line. This is extremely exciting to see how easy it becomes to edit post meta inline 🎉 

 I did a very extensive testing using the admin and author roles, all supported block types: Paragraph, Heading, Button and Image, product and site editor. I couldn't discover any blockers. I would be in favor of including it in WordPress 6.6 release.

SantosGuillamot commented:
Thanks a lot for the extensive testing, Greg! Much appreciated ☺️ 

I totally agree that we will need to keep iterating on the "Enter" and "Paste" workflows, so let's keep an eye on that. Also, once it is out we will get community feedback that can help make decisions.

> I tried to set a binding on the caption attribute, but it is still not covered. However, it is listed in the sidebar.

Regarding this, I just created [this pull request](https://github.com/WordPress/gutenberg/pull/62183) to remove them from the bindings panel.
